ls1intum/Artemis
# Feature/programming exercises/override results

Fixes issue #1162 

### Checklist

### Motivation and Context
Instructors should be able to override programming exercise results
I also refactored some duplicated code and improved the logic on the server so that only tutors can save / submit / override their own assessment before the assessment due date and so that instructors can always do it for modeling, file upload, text and programming exercises

- [x] There is still one TODO with respect to code duplication in my changes, that need to be fixed
- [x] Also, it would be great to explicitly test the new logic with server integration tests for all modeling, file upload, text and programming exercises

### Steps for Testing
1. Add manual result for a programming exercise
2. Instructors should always be able to override it
3. Tutors should not be able to override it
4. Also check this for modeling, file upload and text exercises
